ethier/gen.go

```diff
@@ -99,5 +119,146 @@ func gen(_ *cobra.Command, args []string) (retErr error) {
+					}
+					return parts[1], nil
+				},
+			}).
+			Parse(extraCode),
+	)
+
+	// Regular expressions for modifying abigen-generated code to work with the
+	// extraTemplate code above.
+	deployedRegexp = regexp.MustCompile(`^\s*return address, tx, &(.+?)\{.*Transactor.*\}, nil\s*$`)
+	// Note the option for matching strings.Replace or strings.ReplaceAll due to
+	// a recent change in abigen.
+	libReplacementRegexp = regexp.MustCompile(`^\s*(.+?)Bin = strings.Replace(?:All)?\(.+?, "__\$([0-9a-f]{34})\$__", (.+?)(?:, -1)?\)\s*$`)
+	// TODO(aschlosberg) replace regular expressions with a more explicit
+	// approach for modifying the output code. This likely requires a PR to the
+	// go-ethereum repo to allow bind.Bind (+/- abigen) to accept an alternate
+	// template.
+)
+
+// extendGeneratedCode adds ethier-specific functionality to code generated by
+// abigen, allowing for interoperability with the ethier/solidity package for
+// source-map interpretation at runtime.
+func extendGeneratedCode(generated, combinedJSON *bytes.Buffer) ([]byte, error) {
+	meta := struct {
+		SourceList []string `json:"sourceList"`
+		Version    string   `json:"version"`
+
+		Contracts    map[string]*compiler.Contract
+		CombinedJSON string
+	}{CombinedJSON: combinedJSON.String()}
+
+	if err := json.Unmarshal(combinedJSON.Bytes(), &meta); err != nil {
+		return nil, fmt.Errorf("json.Unmarshal([solc output], %T): %v", &meta, err)
+	}
+
+	cs, err := compiler.ParseCombinedJSON(combinedJSON.Bytes(), "", "", meta.Version, "")
+	if err != nil {
+		return nil, fmt.Errorf("compiler.ParseCombinedJSON(): %v", err)
+	}
+	meta.Contracts = cs
+	for k, c := range meta.Contracts {
+		if c.RuntimeCode == "0x" {
+			delete(meta.Contracts, k)
+		}
+	}
+
+	if err := extraTemplate.Execute(generated, meta); err != nil {
+		return nil, fmt.Errorf("%T.Execute(): %v", extraTemplate, err)
+	}
+
+	// When using vm.Config.Trace, the only contract-identifying information is
+	// the address to which the transaction was sent. We must therefore modify
+	// every DeployFoo() function to save the address(es) at which the contract
+	// is deployed.
+	lines := strings.Split(generated.String(), "\n")
+	for i, l := range lines {
+		matches := deployedRegexp.FindStringSubmatch(l)
+		if len(matches) == 0 {
+			continue
+		}
+		lines[i] = fmt.Sprintf(
+			`deployedContracts[address] = %q // Added by ethier gen
+			%s`,
+			matches[1], l,
+		)
+	}
+
+	// Libraries have their addresses string-replaced directly into contract
+	// code, which we need to mirror for the runtime code too.
+	for i, l := range lines {
+		matches := libReplacementRegexp.FindStringSubmatch(l)
+		if len(matches) == 0 {
+			continue
+		}
+		lines[i] = fmt.Sprintf(
+			`%s
+			RuntimeSourceMaps[%q].RuntimeCode = strings.Replace(RuntimeSourceMaps[%[2]q].RuntimeCode, "__$%s$__", %s, -1)`,
+			l, matches[1], matches[2], matches[3],
+		)
```

Comment:
Oh yeah, this entire implementation is extremely brittle (hence the `--experimental`). I've sent https://github.com/ethereum/go-ethereum/pull/25055 to allow us to completely override the code generation.

Reply:
After rereading the `abigen`-generated code I noticed that it always deploys the same contract bytecode for repeated execution anyways. e.g.
```golang
sourceMapTestLibAddr, _, _, _ := DeploySourceMapTestLib(auth, backend)
SourceMapTest0Bin = strings.Replace(SourceMapTest0Bin, "__$c72c7067445523f34cb493d4791fe09994$__", sourceMapTestLibAddr.String()[2:], -1)
address, tx, contract, err := bind.DeployContract(auth, *parsed, common.FromHex(SourceMapTest0Bin), backend)
```
so even though the library is redeployed, it's address is not actually replaced in the contract bytecode because the pattern no longer applies after the first iteration.

I don't know if this is intended - especially since using `SourceMapTest0Bin` is marked as deprecated.

Reply:
I've been thinking about this bit a lot and I think we can actually just replace the entire `__$(hex)$__` with zeroes, inside the `*solidity.SourceMap` because all we care about is that it can be decoded from hex. My gut tells me that the preceding byte will always be a `PUSH20` for an address.